azure_functions_worker/dispatcher.py

```diff
@@ -485,6 +486,31 @@ async def _handle__function_environment_reload_request(self, req):
                 request_id=self.request_id,
                 function_environment_reload_response=failure_response)
 
+    async def _handle__close_shared_memory_resources_request(self, req):
```

Comment:
One concern. When the Python worker breaks while executing a customer's function, the current behavior is that our host will start a new Python worker. In this scenario, will the share memory region stay opens? Or is there some mechanism from the host to clean this up?

Reply:
My understanding is that in case of Windows, when all processes holding a reference to the memory map go away, the memory map will also be freed.
In Linux, though, since they are files, if a worker dies before it gets the `CloseSharedMemoryResourcesRequest` message then we may have open files. 
Although, note that there are two different cases:
For input bindings, the host will create the map and delete it when the worker sends back a response (even if it was a failure response as you can verify [here](https://github.com/Azure/azure-functions-host/blob/13a43aaa7d7f8441a51233083e692cc584f49143/src/WebJobs.Script.Grpc/Channel/GrpcWorkerChannel.cs#L515)).
For output bindings, the worker will create the map and the host will send back a `CloseSharedMemoryResourcesRequest` message after the host has read the worker's output. So if the worker dies after it had created the map but before it got the message to close it from the host, then we may be in a bad state. Though the fact that the map was created successfully, I think, implies that the execution did succeed so the worker *should* not die.

But yes, in theory, there is a window of time where maps can be left unclosed.
Do you have any proposal on how to counter this?
Maybe we should have some background process clean things up?

Reply:
When starting the host we probably should check the folder and log the error if there are files hanging there.
I would do this at startup of the C# host.

Reply:
@Hazhzeng @vrdmr Is there likely to be a scenario where two host instances are running on the same VM? And they end up sharing `/dev/shm`? In that case, theoretically one process can read content of other applications. Also, since people may know that `/dev/shm` is where we keep these memory maps, people may write functions that try to read content of this directory. Is this a potential security vulnerability? 
cc: @goiri

Reply:
Hi @gohar94, so far in Python, in all SKUs, both the host and worker is bundled in a single container, which means it is very unlikely /dev/shm will be shared among two host instances.

However, a single host can create **multiple worker processes** via `FUNCTIONS_WORKER_PROCESS_COUNT=n` app settings. We just need to double check that in this case, each worker process does not interfere each other.

Reply:
Yeah that should be fine. We have been running our internal version with similar changes with 10 worker processes and things work fine.
Each memory map name is a randomly generated UUID so it's unlikely that there will be collisions or interface.

Reply:
Any clarification on this @gohar94?

Reply:
So the problem is, we can't make the worker remove maps until the host is done reading them. Or unless the host indicates it's safe to do so (for example, if we add caching later, the host might not send this message to remove the map right away.)
Therefore, in my opinion assuming that the host is going to be responsible for closing these (or making the worker do it) is fair.

Reply:
The problem with a periodic check is going to be that if later on, we come up with some policy where the host does not want a particular map to be deleted, then it makes the protocol between host/worker pretty complex.

Reply:
Okay sounds fair - Lets see this in action and keep this in mind.

I agree with @goiri's proposal - if either in the host or in the worker, we should log during startup.

> When starting the host we probably should check the folder and log the error if there are files hanging there.
> I would do this at startup of the C# host.

Reply:
Yes, logging from the host sounds good. If we log from the worker and in the case where there are more than 1 workers, the workers that start up after the first one might see files in those directories (which is completely fine.)